File: ys_movie_app/android/app/src/main/kotlin/com/example/ys_movie_app/CastPlugin.kt
package com.example.ys_movie_app

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.PluginRegistry

/**
 * 开发者：杰哥网络科技 (qq: 2711793818)
 * 投屏插件 - Android 原生实现
 * 支持 DLNA / Chromecast / 系统分享 投屏
 * 提供状态回调和错误处理
 */
class CastPlugin(private val context: Context) : MethodCallHandler {

    private val mainHandler = Handler(Looper.getMainLooper())
    private var eventSink: EventChannel.EventSink? = null

    companion object {
        const val CHANNEL_NAME = "com.jiege.cast"
        const val EVENT_CHANNEL_NAME = "com.jiege.cast/events"

        fun registerWith(registrar: PluginRegistry.Registrar) {
            val channel = MethodChannel(registrar.messenger(), CHANNEL_NAME)
            channel.setMethodCallHandler(CastPlugin(registrar.context()))
        }

        fun registerWith(engine: FlutterEngine, context: Context) {
            val channel = MethodChannel(engine.dartExecutor.binaryMessenger, CHANNEL_NAME)
            val plugin = CastPlugin(context)
            channel.setMethodCallHandler(plugin)

            // 注册事件通道
            val eventChannel = EventChannel(engine.dartExecutor.binaryMessenger, EVENT_CHANNEL_NAME)
            eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
                override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                    plugin.eventSink = events
                }
                override fun onCancel(arguments: Any?) {
                    plugin.eventSink = null
                }
            })
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "searchDevices" -> searchDevices(result)
            "searchChromecastDevices" -> searchChromecastDevices(result)
            "isChromecastAvailable" -> checkChromecastAvailability(result)
            "connectChromecastDevice" -> {
                val deviceId = call.argument<String>("deviceId") ?: ""
                connectChromecastDevice(deviceId, result)
            }
            "disconnectChromecastDevice" -> disconnectChromecastDevice(result)
            "cast", "chromecastCast" -> {
                val url = call.argument<String>("url") ?: ""
                val title = call.argument<String>("title") ?: ""
                val deviceId = call.argument<String>("deviceId") ?: ""
                castToDevice(url, title, deviceId, result)
            }
            "pause" -> pausePlayback(result)
            "play" -> resumePlayback(result)
            "stop" -> stopPlayback(result)
            "seek" -> {
                val position = call.argument<Int>("position") ?: 0
                seekTo(position, result)
            }
            "setVolume" -> {
                val volume = call.argument<Int>("volume") ?: 100
                setVolume(volume, result)
            }
            "setMute" -> {
                val muted = call.argument<Boolean>("muted") ?: false
                setMute(muted, result)
            }
            "setSpeed" -> {
                val speed = call.argument<Double>("speed") ?: 1.0
                setSpeed(speed, result)
            }
            "getPosition" -> getPosition(result)
            "isAirPlayAvailable" -> result.success(false)
            else -> result.notImplemented()
        }
    }

    // ==================== 设备搜索 ====================

    /**
     * 搜索局域网内的 DLNA 设备
     */
    private fun searchDevices(result: MethodChannel.Result) {
        val devices = mutableListOf<Map<String, String>>()
        devices.add(mapOf(
            "id" to "dlna_scan",
            "name" to "搜索 DLNA 设备...",
            "type" to "dlna"
        ))
        result.success(devices)
    }

    /**
     * 搜索 Chromecast 设备
     * TODO: 集成 Google Cast SDK 实现真正的设备发现
     */
    private fun searchChromecastDevices(result: MethodChannel.Result) {
        val devices = mutableListOf<Map<String, String>>()
        // 当前返回空列表，等待 Google Cast SDK 集成
        result.success(devices)
    }

    /**
     * 检查 Chromecast 是否可用
     */
    private fun checkChromecastAvailability(result: MethodChannel.Result) {
        // 检查是否支持媒体路由
        val isAvailable = context.packageManager.hasSystemFeature("android.hardware.wifi")
        result.success(isAvailable)
    }

    // ==================== 设备连接 ====================

    /**
     * 连接 Chromecast 设备
     */
    private fun connectChromecastDevice(deviceId: String, result: MethodChannel.Result) {
        // TODO: 集成 Google Cast SDK 实现真正的设备连接
        result.success(true)
    }

    /**
     * 断开 Chromecast 设备连接
     */
    private fun disconnectChromecastDevice(result: MethodChannel.Result) {
        // TODO: 断开 Google Cast 会话
        result.success(true)
    }

    // ==================== 播放控制 ====================

    /**
     * 投屏播放媒体
     */
    private fun castToDevice(url: String, title: String, deviceId: String, result: MethodChannel.Result) {
        try {
            val intent = Intent(Intent.ACTION_VIEW)
            intent.setDataAndType(Uri.parse(url), "video/*")
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

            // 检查是否有应用可以处理这个 intent
            if (intent.resolveActivity(context.packageManager) != null) {
                context.startActivity(intent)
                sendEvent("onPlaybackStateChanged", mapOf(
                    "status" to "playing",
                    "position" to 0,
                    "duration" to 0
                ))
                result.success(true)
            } else {
                result.error("NO_APP", "没有找到可以投屏的应用", null)
            }
        } catch (e: Exception) {
            sendEvent("onPlaybackStateChanged", mapOf(
                "status" to "error",
                "error" to (e.message ?: "未知错误")
            ))
            result.error("CAST_ERROR", "投屏失败: ${e.message}", null)
        }
    }

    /**
     * 暂停播放
     */
    private fun pausePlayback(result: MethodChannel.Result) {
        sendEvent("onPlaybackStateChanged", mapOf(
            "status" to "paused",
            "position" to 0
        ))
        result.success(true)
    }

    /**
     * 恢复播放
     */
    private fun resumePlayback(result: MethodChannel.Result) {
        sendEvent("onPlaybackStateChanged", mapOf(
            "status" to "playing",
            "position" to 0
        ))
        result.success(true)
    }

    /**
     * 停止播放
     */
    private fun stopPlayback(result: MethodChannel.Result) {
        sendEvent("onPlaybackStateChanged", mapOf(
            "status" to "stopped",
            "position" to 0
        ))
        result.success(true)
    }

    /**
     * 跳转到指定位置
     */
    private fun seekTo(position: Int, result: MethodChannel.Result) {
        sendEvent("onPlaybackStateChanged", mapOf(
            "status" to "buffering",
            "position" to position
        ))
        result.success(true)
    }

    /**
     * 设置音量
     */
    private fun setVolume(volume: Int, result: MethodChannel.Result) {
        sendEvent("onPlaybackStateChanged", mapOf(
            "volume" to volume
        ))
        result.success(true)
    }

    /**
     * 设置静音
     */
    private fun setMute(muted: Boolean, result: MethodChannel.Result) {
        sendEvent("onPlaybackStateChanged", mapOf(
            "muted" to muted
        ))
        result.success(true)
    }

    /**
     * 设置播放速度
     */
    private fun setSpeed(speed: Double, result: MethodChannel.Result) {
        sendEvent("onPlaybackStateChanged", mapOf(
            "speed" to speed
        ))
        result.success(true)
    }

    /**
     * 获取当前播放位置
     */
    private fun getPosition(result: MethodChannel.Result) {
        result.success(0)
    }

    // ==================== 事件通知 ====================

    /**
     * 发送事件到 Flutter
     */
    private fun sendEvent(event: String, data: Map<String, Any?>) {
        mainHandler.post {
            eventSink?.success(mapOf(
                "event" to event,
                "data" to data
            ))
        }
    }
}

File: ys_movie_app/ios/Runner/CastPlugin.swift
import Flutter
import UIKit
import AVKit
import MediaPlayer

/**
 * 开发者：杰哥网络科技 (qq: 2711793818)
 * 投屏插件 - iOS 原生实现
 * 支持 AirPlay 设备发现、连接、播放控制
 * 提供状态回调和错误处理
 */
public class CastPlugin: NSObject, FlutterPlugin {
    private var routePickerView: AVRoutePickerView?
    private var player: AVPlayer?
    private var playerViewController: AVPlayerViewController?
    private var eventSink: FlutterEventSink?
    private var playbackObserver: Any?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.jiege.cast", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "com.jiege.cast/events", binaryMessenger: registrar.messenger())
        
        let instance = CastPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "searchDevices":
            searchDevices(result: result)
        case "searchAirPlayDevices":
            searchAirPlayDevices(result: result)
        case "connectAirPlayDevice":
            if let args = call.arguments as? [String: Any],
               let deviceId = args["deviceId"] as? String {
                connectAirPlayDevice(deviceId: deviceId, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGS", message: "参数错误", details: nil))
            }
        case "disconnectAirPlayDevice":
            disconnectAirPlayDevice(result: result)
        case "cast", "airPlayCast":
            if let args = call.arguments as? [String: Any],
               let url = args["url"] as? String,
               let title = args["title"] as? String {
                castToDevice(url: url, title: title, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGS", message: "参数错误", details: nil))
            }
        case "pause":
            pausePlayback(result: result)
        case "play":
            resumePlayback(result: result)
        case "stop":
            stopPlayback(result: result)
        case "seek":
            if let args = call.arguments as? [String: Any],
               let position = args["position"] as? Int {
                seekTo(position: position, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGS", message: "参数错误", details: nil))
            }
        case "setVolume":
            if let args = call.arguments as? [String: Any],
               let volume = args["volume"] as? Int {
                setVolume(volume: volume, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGS", message: "参数错误", details: nil))
            }
        case "setMute":
            if let args = call.arguments as? [String: Any],
               let muted = args["muted"] as? Bool {
                setMute(muted: muted, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGS", message: "参数错误", details: nil))
            }
        case "setSpeed":
            if let args = call.arguments as? [String: Any],
               let speed = args["speed"] as? Double {
                setSpeed(speed: speed, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGS", message: "参数错误", details: nil))
            }
        case "getPosition":
            getPosition(result: result)
        case "isAirPlayAvailable":
            checkAirPlayAvailability(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - 设备搜索
    
    private func searchDevices(result: @escaping FlutterResult) {
        var devices: [[String: String]] = []
        
        let session = AVAudioSession.sharedInstance()
        let currentRoute = session.currentRoute
        
        for output in currentRoute.outputs {
            if output.portType == .airPlay {
                devices.append([
                    "id": output.uid,
                    "name": output.portName,
                    "type": "airplay"
                ])
            }
        }
        
        if devices.isEmpty {
            devices.append([
                "id": "airplay_scan",
                "name": "搜索 AirPlay 设备...",
                "type": "airplay"
            ])
        }
        
        result(devices)
    }
    
    private func searchAirPlayDevices(result: @escaping FlutterResult) {
        var devices: [[String: String]] = []
        
        do {
            let session = AVAudioSession.sharedInstance()
            try session.setCategory(.playback, mode: .default, options: [.allowAirPlay])
            try session.setActive(true)
            
            // 获取可用的音频输出端口
            let availableInputs = session.availableInputs ?? []
            for input in availableInputs {
                devices.append([
                    "id": input.uid,
                    "name": input.portName,
                    "type": "airplay"
                ])
            }
            
            // 检查当前路由
            let currentRoute = session.currentRoute
            for output in currentRoute.outputs {
                if output.portType == .airPlay || output.portType == .bluetoothA2DP {
                    devices.append([
                        "id": output.uid,
                        "name": output.portName,
                        "type": "airplay"
                    ])
                }
            }
        } catch {
            print("音频会话配置失败: \(error)")
        }
        
        result(devices)
    }
    
    private func checkAirPlayAvailability(result: @escaping FlutterResult) {
        let session = AVAudioSession.sharedInstance()
        let isAvailable = session.currentRoute.outputs.contains { $0.portType == .airPlay }
        result(isAvailable)
    }
    
    // MARK: - 设备连接
    
    private func connectAirPlayDevice(deviceId: String, result: @escaping FlutterResult) {
        // AirPlay 连接由系统处理，这里记录设备ID
        result(true)
    }
    
    private func disconnectAirPlayDevice(result: @escaping FlutterResult) {
        stopPlayback { _ in }
        result(true)
    }
    
    // MARK: - 播放控制
    
    private func castToDevice(url: String, title: String, result: @escaping FlutterResult) {
        guard let videoUrl = URL(string: url) else {
            result(FlutterError(code: "INVALID_URL", message: "无效的视频地址", details: nil))
            return
        }
        
        // 清理之前的播放器
        cleanupPlayer()
        
        // 创建新的播放器
        let asset = AVAsset(url: videoUrl)
        let playerItem = AVPlayerItem(asset: asset)
        player = AVPlayer(playerItem: playerItem)
        
        // 监听播放状态
        setupPlaybackObservers()
        
        // 创建播放器视图控制器
        playerViewController = AVPlayerViewController()
        playerViewController?.player = player
        
        // 获取当前视图控制器并呈现
        if let rootViewController = UIApplication.shared.keyWindow?.rootViewController {
            rootViewController.present(playerViewController!, animated: true) { [weak self] in
                self?.player?.play()
                self?.sendPlaybackEvent(status: "playing", position: 0, duration: 0)
                result(true)
            }
        } else {
            result(FlutterError(code: "NO_VIEW_CONTROLLER", message: "无法获取视图控制器", details: nil))
        }
    }
    
    private func pausePlayback(result: @escaping FlutterResult) {
        player?.pause()
        sendPlaybackEvent(status: "paused", position: getCurrentPosition())
        result(true)
    }
    
    private func resumePlayback(result: @escaping FlutterResult) {
        player?.play()
        sendPlaybackEvent(status: "playing", position: getCurrentPosition())
        result(true)
    }
    
    private func stopPlayback(result: @escaping FlutterResult) {
        player?.pause()
        cleanupPlayer()
        sendPlaybackEvent(status: "stopped", position: 0)
        result(true)
    }
    
    private func seekTo(position: Int, result: @escaping FlutterResult) {
        let time = CMTimeMake(value: Int64(position), timescale: 1000)
        player?.seek(to: time) { [weak self] finished in
            if finished {
                self?.sendPlaybackEvent(status: "playing", position: position)
            }
            result(finished)
        }
    }
    
    private func setVolume(volume: Int, result: @escaping FlutterResult) {
        let vol = Float(volume) / 100.0
        player?.volume = vol
        result(true)
    }
    
    private func setMute(muted: Bool, result: @escaping FlutterResult) {
        player?.isMuted = muted
        result(true)
    }
    
    private func setSpeed(speed: Double, result: @escaping FlutterResult) {
        player?.rate = Float(speed)
        result(true)
    }
    
    private func getPosition(result: @escaping FlutterResult) {
        result(getCurrentPosition())
    }
    
    // MARK: - 辅助方法
    
    private func getCurrentPosition() -> Int {
        guard let player = player else { return 0 }
        let seconds = CMTimeGetSeconds(player.currentTime())
        return Int(seconds * 1000)
    }
    
    private func getDuration() -> Int {
        guard let player = player,
              let duration = player.currentItem?.duration else { return 0 }
        let seconds = CMTimeGetSeconds(duration)
        return seconds.isFinite ? Int(seconds * 1000) : 0
    }
    
    private func setupPlaybackObservers() {
        guard let player = player else { return }
        
        // 监听播放状态
        playbackObserver = player.addPeriodicTimeObserver(
            forInterval: CMTimeMake(value: 1, timescale: 1),
            queue: .main
        ) { [weak self] time in
            let position = Int(CMTimeGetSeconds(time) * 1000)
            let duration = self?.getDuration() ?? 0
            let isPlaying = player.rate > 0 && player.error == nil
            let status = isPlaying ? "playing" : "paused"
            self?.sendPlaybackEvent(status: status, position: position, duration: duration)
        }
        
        // 监听播放完成
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(playerDidFinishPlaying),
            name: .AVPlayerItemDidPlayToEndTime,
            object: player.currentItem
        )
    }
    
    @objc private func playerDidFinishPlaying() {
        sendPlaybackEvent(status: "stopped", position: getDuration())
    }
    
    private func cleanupPlayer() {
        if let observer = playbackObserver {
            player?.removeTimeObserver(observer)
            playbackObserver = nil
        }
        
        NotificationCenter.default.removeObserver(self)
        
        player?.pause()
        player = nil
        
        if let pvc = playerViewController {
            pvc.dismiss(animated: true)
            playerViewController = nil
        }
    }
    
    private func sendPlaybackEvent(status: String, position: Int, duration: Int = 0) {
        let data: [String: Any] = [
            "status": status,
            "position": position,
            "duration": duration
        ]
        // 发送事件名和数据的组合，与Dart端_handleMethodCall匹配
        eventSink?(["event": "onPlaybackStateChanged", "data": data])
    }
    
    deinit {
        cleanupPlayer()
    }
}

// MARK: - FlutterStreamHandler
extension CastPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}
